src/main/java/dev/morling/onebrc/CalculateAverage_royvanrijn.java

```diff
@@ -16,64 +16,224 @@
+            if (newline == '\r')
+                bb.get();
+
+            // Update the map, computeIfAbsent has the least amount of branches I think, compared to get()/put() or merge() or compute():
+            Measurement measurement = measurements.computeIfAbsent(key, (k) -> new Measurement());
+            measurement.updateWith(measured);
         }
+
+        return measurements;
     }
 
-    public static void main(String[] args) throws IOException {
+    /**
+     * Thanks to bjhara for the idea of using memory mapped files, TIL.
+     * @param fileChannel
+     * @return
+     * @throws IOException
+     */
+    private static Stream<ByteBuffer> splitFileChannel(final FileChannel fileChannel) throws IOException {
+        return StreamSupport.stream(Spliterators.spliteratorUnknownSize(new Iterator<>() {
+            private static final long CHUNK_SIZE = (long) Math.pow(2, 19);
+
+            private final long size = fileChannel.size();
+            private long bytesRead = 0;
+
+            @Override
+            public ByteBuffer next() {
+                try {
+                    MappedByteBuffer mappedByteBuffer = fileChannel.map(FileChannel.MapMode.READ_ONLY, bytesRead, Math.min(CHUNK_SIZE, size - bytesRead));
+
+                    int realEnd = mappedByteBuffer.limit() - 1;
+                    while (mappedByteBuffer.get(realEnd) != '\n') {
+                        realEnd--;
+                    }
+                    mappedByteBuffer.limit(++realEnd);
+                    bytesRead += realEnd;
 
-//        long before = System.currentTimeMillis();
+                    return mappedByteBuffer;
+                }
+                catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+            }
 
-        Map<String, Measurement> resultMap = Files.lines(Path.of(FILE)).parallel()
-                .map(record -> {
-                    // Map to <String,double>
-                    int pivot = record.indexOf(";");
-                    String key = record.substring(0, pivot);
-                    double measured = Double.parseDouble(record.substring(pivot + 1));
-                    return new AbstractMap.SimpleEntry<>(key, measured);
-                })
-                .collect(Collectors.toConcurrentMap(
-                        // Combine/reduce:
-                        AbstractMap.SimpleEntry::getKey,
-                        entry -> new Measurement(entry.getValue()),
-                        Measurement::combineWith));
+            @Override
+            public boolean hasNext() {
+                return bytesRead < size;
+            }
+        }, Spliterator.IMMUTABLE), false);
+    }
+
+    /**
+     * Branchless parser, goes from String to int (10x):
+     * "-1.2" to -12
+     * "40.1" to 401
+     * etc.
+     *
+     * @param input
+     * @return int value x10
+     */
+    private static int branchlessParseInt(final byte[] input, int length) {
```

Comment:
I wonder if you would get any performance improvements caching the temperature to int? Assuming temperature ranges from -99.9 to 99.9, there would only be at most ~2000 keys?